cmd/gardenlet/app/gardenlet.go

```diff
@@ -0,0 +1,462 @@
+	data, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+	return o.decodeConfig(data)
+}
+
+// decodeConfig decodes data as a GardenletConfiguration object.
+func (o *Options) decodeConfig(data []byte) (*config.GardenletConfiguration, error) {
+	gardenletConfig := &config.GardenletConfiguration{}
+	if _, _, err := o.codecs.UniversalDecoder().Decode(data, nil, gardenletConfig); err != nil {
+		return nil, err
+	}
+	return gardenletConfig, nil
+}
+
+func (o *Options) configFileSpecified() error {
+	if len(o.ConfigFile) == 0 {
+		return fmt.Errorf("missing Gardenlet config file")
+	}
+	return nil
+}
+
+// Validate validates all the required options.
+func (o *Options) validate(args []string) error {
+	if len(args) != 0 {
+		return errors.New("arguments are not supported")
+	}
+
+	return nil
+}
+
+func (o *Options) applyDefaults(in *config.GardenletConfiguration) (*config.GardenletConfiguration, error) {
+	external, err := o.scheme.ConvertToVersion(in, configv1alpha1.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	o.scheme.Default(external)
+
+	internal, err := o.scheme.ConvertToVersion(external, config.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	out := internal.(*config.GardenletConfiguration)
+
+	return out, nil
+}
+
+func (o *Options) run(ctx context.Context) error {
+	if len(o.ConfigFile) > 0 {
+		c, err := o.loadConfigFromFile(o.ConfigFile)
+		if err != nil {
+			return err
+		}
+
+		c, err = o.applyDefaults(c)
+		if err != nil {
+			return err
+		}
+
+		if errs := configvalidation.ValidateGardenletConfiguration(c); len(errs) > 0 {
+			return fmt.Errorf("errors validating the configuration: %+v", errs)
+		}
+
+		o.config = c
+	}
+
+	// Add feature flags
+	if err := features.FeatureGate.SetFromMap(o.config.FeatureGates); err != nil {
+		return err
+	}
+
+	gardenlet, err := NewGardenlet(ctx, o.config)
+	if err != nil {
+		return err
+	}
+
+	return gardenlet.Run(ctx)
+}
+
+// NewCommandStartGardenlet creates a *cobra.Command object with default parameters
+func NewCommandStartGardenlet(ctx context.Context) *cobra.Command {
+	opts, err := NewOptions()
+	if err != nil {
+		panic(err)
+	}
+
+	cmd := &cobra.Command{
+		Use:   "gardenlet",
+		Short: "Launch the Gardenlet",
+		Long: `In essence, the Gardener is an extension API server along with a bundle
+of Kubernetes controllers which introduce new API objects in an existing Kubernetes
+cluster (which is called Garden cluster) in order to use them for the management of
+further Kubernetes clusters (which are called Shoot clusters).
+To do that reliably and to offer a certain quality of service, it requires to control
+the main components of a Kubernetes cluster (etcd, API server, controller manager, scheduler).
+These so-called control plane components are hosted in Kubernetes clusters themselves
+(which are called Seed clusters).`,
+		Run: func(cmd *cobra.Command, args []string) {
```

Comment:
```suggestion
		RunE: func(cmd *cobra.Command, args []string) error {
```

I think RunE improves the testability and instead of panics we ca just return the error. The `err` will be propagated to the invocation of `command.Execute()`.